### PR TITLE
[JBPM-9572] Duplicate process instances (with timer) created after up…

### PR DESCRIPTION
**Pull request by elguardian**
Jira: https://issues.redhat.com/browse/JBPM-9572

### CODE REVIEW

File: jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java [hunk 1]
```diff
@@ -24,7 +24,10 @@
 import java.io.Serializable;
 import java.time.ZonedDateTime;
 import java.time.temporal.ChronoUnit;
+import java.util.Arrays;
```
Comment (gmunozfe) [inline jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java:27]: Not used

File: jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java [hunk 2]
```diff
@@ -24,7 +24,10 @@
...
 import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
```
Comment (gmunozfe) [inline jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java:30]: Not used

File: jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java [hunk 3]
```diff
@@ -43,12 +46,15 @@
 import javax.ejb.TimerConfig;
 import javax.ejb.TransactionManagement;
 import javax.ejb.TransactionManagementType;
+import javax.transaction.RollbackException;
 import javax.transaction.Status;
 import javax.transaction.UserTransaction;
 
 import org.drools.core.time.JobHandle;
 import org.drools.core.time.impl.TimerJobInstance;
+import org.drools.persistence.api.TransactionManagerHelper;
```
Comment (gmunozfe) [inline jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java:55]: Not used

File: jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java [hunk 4]
```diff
@@ -43,12 +46,15 @@
...
 import org.jbpm.process.core.timer.TimerServiceRegistry;
+import org.jbpm.process.instance.timer.TimerManager.StartProcessJob;
```
Comment (gmunozfe) [inline jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java:57]: Not used

File: jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java [hunk 5]
```diff
@@ -84,6 +90,9 @@
     @Resource
     protected UserTransaction utx;
 
+    @Resource
+    protected javax.transaction.TransactionSynchronizationRegistry tsr;
```
Comment (gmunozfe) [inline jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java:94]: Not used, is it needed?

File: jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java [hunk 6]
```diff
@@ -119,7 +128,7 @@ public void executeTimerJob(Timer timer) {
         try {
             transaction(this::executeTimerJobInstance, timerJobInstance);
         } catch (Exception e) {
-            recoverTimerJobInstance(timerJob, e);
+            recoverTimerJobInstance(timer, timerJob, e);
```
Comment (gmunozfe) [inline jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java:131]: timer is not used in recoverTimeJobInstance

File: jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java [hunk 7]
```diff
@@ -131,13 +140,18 @@ private void executeTimerJobInstance(TimerJobInstance timerJobInstance) throws E
         }
     }
 
-    private void recoverTimerJobInstance(EjbTimerJob ejbTimerJob, Exception e) {
+    private void recoverTimerJobInstance(Timer timer, EjbTimerJob ejbTimerJob, Exception e) {
```
Comment (gmunozfe) [inline jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java:143]: timer is not used in recoverTimeJobInstance

**Review by gmunozfe — COMMENTED** [review]
Looks good, just remove some unused imports and newly added param to recover method

File: jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java (hunk 1 above)
Comment (elguardian) [reply jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java:27]: removed

File: jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java (hunk 3 above)
Comment (elguardian) [reply jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java:55]: removed

File: jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java (hunk 4 above)
Comment (elguardian) [reply jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java:57]: removed

File: jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java (hunk 5 above)
Comment (elguardian) [reply jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java:94]: removed

File: jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java (hunk 6 above)
Comment (elguardian) [reply jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java:131]: removed

File: jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java (hunk 7 above)
Comment (elguardian) [reply jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java:143]: removed

**Review by gmunozfe — APPROVED** [review]
Looks good to me, brilliant work @elguardian !

**Pull request merged**